### PR TITLE
feat: nolocal flag support for bridges

### PR DESCRIPTION
**Pull request by oldrich-s**
https://github.com/moscajs/aedes/issues/736

### CODE REVIEW

File: lib/handlers/subscribe.js [hunk 1]
```diff
@@ -142,13 +142,12 @@ function addSubs (sub, done) {
...
-      _packet = new Packet(_packet, broker)
-      _packet.retain = false
-      deliverFunc(_packet, cb)
-    }
+  const deliverFunc = func
+  func = function handlePacketSubscription (_packet, cb) {
+    _packet = new Packet(_packet, broker)
+    _packet.nl = nl
+    if (!rap) _packet.retain = false
+    deliverFunc(_packet, cb)
```
Comment (robertsLando) [inline lib/handlers/subscribe.js:150]: Could you explain this?

File: lib/client.js [hunk 2]
```diff
@@ -116,7 +119,10 @@ function Client (broker, conn, req) {
     }
     const toForward = dedupe(that, _packet) &&
       that.broker.authorizeForward(that, _packet)
-    if (toForward) {
+
+    const isNL = _packet.clientId === that.id && _packet.nl
+
+    if (toForward && !isNL) {
```
Comment (robertsLando) [inline lib/client.js:125]: Could you create a `shouldForward` method for this as it's the same for deliver0 and deliverQos?

**Review by robertsLando — CHANGES_REQUESTED** [review]

**Comment by robertsLando** [comment]
Also please add a unit test that covers this

**Comment by robertsLando** [comment]
@gnought Opinions on this one?

File: lib/handlers/subscribe.js (hunk 1 above)
Comment (oldrich-s) [reply lib/handlers/subscribe.js:150]: Originally, the code checked if rap (retain as published) was `false` (which is default) and if so then it wrapped the delivery function and injected / modified the retain flag to `false` to indicate that the message was not retained.

I have made a modification that always wraps the delivery function and always injects nl (nolocal) flag from the subscription into the packet. The `func` is then sent to mqemmiter that uses it whenever a packet is received.

So this way I make the `nl` flag from the subscription accessible in the `deliver0` and `deliverQoS` in the client.js.

File: lib/client.js (hunk 2 above)
Comment (oldrich-s) [reply lib/client.js]: Should be done now 👍

File: test/bridge.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  })
+})
+
+test('bridge client sends a publish message but shall not receive it back', function (t) {
+  const s = connect(setup(), { clientId: 'my-client-bridge-1', protocolVersion: 128 + 4 })
+  t.teardown(s.broker.close.bind(s.broker))
+
+  const handle = setTimeout(() => t.end(), 1000)
+
+  subscribe(t, s, 'hello', 0, function () {
```
Comment (robertsLando) [inline test/bridge.js:35]: Where do you set the no local flag here?

Comment (oldrich-s) [reply test/bridge.js]: It is set implicitly by mqtt-packet by setting `protocolVersion: 128 + 4`

**Comment by oldrich-s** [comment]
> LGTM. Thanks for this PR @oldrich-s ! Is it ready to merge?

I have just added qos 1 test (just to be sure that persistence works as expected). But otherwise I believe it is ready to be merged.

**Review by robertsLando — CHANGES_REQUESTED** [review]

File: lib/client.js [hunk 4]
```diff
@@ -89,9 +89,18 @@ function Client (broker, conn, req) {
   conn.on('end', this.close.bind(this))
   this._eos = eos(this.conn, this.close.bind(this))
 
-  this.deliver0 = function deliverQoS0 (_packet, cb) {
+  const getToForwardPacket = (_packet) => {
+    if (_packet.clientId === that.id && _packet.nl) return
```
Comment (robertsLando) [inline lib/client.js:93]: Could you add a reference to MQTTv5 Section here? 
```suggestion
    // Mqttv5 3.8.3.1: https://docs.oasis-open.org/mqtt/mqtt/v5.0/mqtt-v5.0.html#_Toc3901169
    // prevent to forward messages sent by the same client when no-local flag is set
    if (_packet.clientId === that.id && _packet.nl) return
```

File: test/bridge.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+          t.end()
+        }
+      })
+  
+      s.inStream.write(packet)
+    })
+  })
+
+  test('bridge client sends a publish message but shall not receive it back, qos = ' + qos, function (t) {
+    const s = connect(setup(), { clientId: 'my-client-bridge-1', protocolVersion: 128 + 4 })
```
Comment (robertsLando) [inline test/bridge.js:38]:
```suggestion
    // protocolVersion 128 + 4 means mqtt 3.1.1 with bridgeMode enabled
    // https://github.com/mqttjs/mqtt-packet/blob/7f7c2ed8bcb4b2c582851d120a94e0b4a731f661/parser.js#L171
    const s = connect(setup(), { clientId: 'my-client-bridge-1', protocolVersion: 128 + 4 })
```

**Review by robertsLando — CHANGES_REQUESTED** [review]

**Comment by oldrich-s** [comment]
> LGTM. Good to merge?

yeps 👍

**Pull request merged**